### PR TITLE
#165305774 Highlight Text and Comment on Articles

### PR DESCRIPTION
**Pull request by matthenge**
### What does this PR do?
- Adds highlight text and comment functionality on articles 
### What are the tasks to be completed?
* Post highlight
* Get all user highlights for article
* Update highlight
* Delete highlight
* Tests
### Local Development Setup
 - Refer to the [README](https://github.com/andela/ah-the-immortals-backend) document for local setup
 - Clone the repo [here](https://github.com/andela/ah-the-immortals-backend.git) 

a. **After cloning the repo:-**
```
$ git checkout  ft-highlight-text-165305774
$ python manage.py runserver
```
b. **Test with postman**
1. After signing up a user, using postman request:
```
POST http://127.0.0.1:8000/api/articles/<slug>/highlight/
GET http://127.0.0.1:8000/api/articles/<slug>/highlight/
PATCH http://127.0.0.1:8000/api/articles/<slug>/highlight/<highlight_id>/
DELETE http://127.0.0.1:8000/api/articles/<slug>/highlight/<highlight_id>/
```

#### Pivotal Tracker
[#165305774](https://www.pivotaltracker.com/story/show/165305774)

#### Screenshots
image

image

image

image

### CODE REVIEW

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]
Hello @matthenge ,
Currently, a user can highlight the same text infinite times, you could limit this to only a single highlight(per text index range) to avoid redundancy in our database:
image

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]
Hello @matthenge ,
If a user highlights a text on the **body** of an article when updating the same highlight, logically they should not update to another section like **title** or **description**. 
image
image


In addition, you could use the text indices as references i.e When updating:
```
1. You could increase or decrease the end index of the same text in the same field
2. You could increase or decrease the start index of the same text in the same field
You can validate them to valid indices
```

File: authors/apps/highlights/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    status_code = 400
+    default_detail = "end_index cannot be greater than field length"
+
+
+class NotPermittedError(APIException):
+    """
+    Exception for no permission to perform an action
+    """
+    status_code = 400
+    default_detail = "You cannot delete or update hightlight that is noy yours"
```
Comment (Ogutu-Brian) [inline authors/apps/highlights/exceptions.py]: Hi @matthenge ,
Please fix the spelling mistake in the exception `You cannot delete or update hightlight that is noy yours`
to You cannot delete or update **highlight** that is **not** yours.

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]

File: authors/apps/highlights/exceptions.py (hunk 1 above)
Comment (matthenge) [reply authors/apps/highlights/exceptions.py]: Already fixed. Thanks for catching that

**Pull request merged**